Add "parent" and "ancestors" predicates

mm-- opened the pull request:
Hi!

First of all, terrific package! Not only does `org-ql`'s clean sexp syntax make it quick an easy to write sophisticated agenda views (whereas before you had to write complicated "skip functions"), the query caching makes rebuilding agendas super fast! I think you've saved me from declaring org-mode bankruptcy. :P

I noticed the `children` and `descendants` predicates didn't seem to have analog `parent` and `ancestors` predicates. So I wrote those, and then later realized you had a WIP branch. D'oh.

Anyway, I think having `parent` and `ancestors` predicates would be super helpful.

For example, I use a non-inherited "project" tag (in `org-tags-exclude-from-inheritance`) to denote projects (so I can't use the workaround in #21). With "ancestors", I can write the following query to find unscheduled "NEXT" items in stuck projects:

```elisp
(org-ql-search (org-agenda-files)
  '(and (todo "NEXT")
	(not (scheduled))
	(ancestors (and (tags-local "project")
			(not (descendants (and (todo)
					       (scheduled)))))))
  :title "Next tasks that should be scheduled")
```

I haven't extensively profiled `ancestors`. You mentioned in your branch that there may be some performance concerns, especially since the query inside of `ancestors` might be tested on siblings (those this is limited with `level`, it seems like it might traverse all siblings [I'm not sure]). But it seems pretty usable to me for now.

I've also added some documentation for the predicates. I haven't included the updated `org-ql.info`, as my version of `makeinfo` seems to be giving different formatting.

alphapapa commented:
Hi Joshua,

Thanks for the kind words.  I'm glad to hear it!  That's what I was aiming for.  I think many Org users would benefit from it, so I try to spread the word about it when I see opportunities.  But, like many Emacs users, Org users can be *reasonably* stubborn--myself included, because we like things to keep working--so progress on this front is slow.  :)

Have you compared the code in the WIP branch to the code in this PR?  I think the simpler code in the WIP branch should return the same results and should run faster.

However, I'm still not satisfied with its implementation, which is why I haven't merged it yet.  The problem is that it repeats all of the work done by calling `org-ql-search` for each ancestor/parent heading, which includes a lot of setup, binding function symbols, byte-compiling the predicate function, etc.  Instead, the code that's run for each ancestor/parent heading should be minimal, and any setup or pre-processing should be done once.

So, for example, we should prepare the ancestor/parent query predicate function ahead of time, preferably in `--pre-process-query`, and the main query's ancestor/parent predicate should take the prepared (i.e. byte-compiled) predicate as its argument.  Then the ancestor/parent predicate should simply go up the outline testing the predicate, returning non-nil when it finds a match.

That sounds almost too simple to me, so, since I haven't looked at that branch in a while, I may have forgotten some details that need to be addressed.  Alternatively, maybe I finally figured out the solution!  :)

mm-- commented:
Ah, actually your code there is smarter about limiting the query space by using `org-entry-end-position` instead of searching the whole subtree.

I think one issue with your code, though, is you can't have something like `(parent (children (tags "something")))`, because when `children` calls `org-narrow-to-subtree`, the heading will already have been narrowed to `(org-entry-end-position)` (and so the children will not be visible). For example, this query for your `parent` selector fails:

```elisp
;; This should just return all TODOs with a parent
(org-ql-search (current-buffer)
  '(and (todo "TODO")
	(parent (children (todo "TODO")))))
```

This is just fixed by adding a `(widen)` before `(org-narrow-to-subtree)` in `children` and `descendants` (although I don't know if this causes issues with other predicates).

Anyway, it sounds like you figured something out, haha. I was looking more through the code (and wondering what a "preamble" was before seeing your comment about naming, haha) and saw all the repeated work you mentioned.

It seems you could have something like this:
```elisp
(org-ql--defpred ancestors (query)
  "Return non-nil if any of current entry's ancestors matches QUERY."
  (save-excursion
    (save-restriction
      (widen)
      ;; Maybe org-ql--pre-process-query can pass this stuff to us
      ;; instead of having to calculate it each time.
      ;; The pcase could have `((ancestors query) (ancestors-preprocessed ))'
      (-let* ((query (org-ql--pre-process-query query))
	      ((&plist :query :preamble :preamble-case-fold) (org-ql--query-preamble query))
	      (predicate (org-ql--query-predicate query))
	      ;; I'm not sure how long byte compiling takes, or if
	      ;; this should be stored elsewhere outside the query.
	      (action (byte-compile `(lambda (&rest _ignore) t))))
	(cl-loop while (org-up-heading-safe)
		 thereis (save-restriction
			   (narrow-to-region (point) (org-entry-end-position))
			   (org-ql--select-cached :query query :preamble preamble :preamble-case-fold preamble-case-fold
						  :predicate predicate :action action :narrow t)))))))
```

But while you wouldn't have to recompile the query while stepping up through ancestors, you do end up byte-compiling the nested query a lot. Then I looked to see if I could do the byte-compiling in `org-ql--pre-process-query`, and then saw that you had already mentioned that in your reply. D'oh, again! It seems like I'm basically retracing your steps, though I haven't grokked the "`org-ql--value-at`" part yet.

I was trying to have something like

```elisp
(defun org-ql--pre-process-ancestors-query (preprocessedquery)
  "Take in a PREPROCESSEDQUERY (or \"(rec query)\" in `org-ql--pre-process-query'
 meant for ancestors, and return a plist suitable to give to `org-ql--select-cached'
`org-ql--select-cached'."
  (-let* (((&plist :query :preamble :preamble-case-fold) (org-ql--query-preamble preprocessedquery))
	  (predicate (org-ql--query-predicate query))
	  (action (byte-compile `(lambda (&rest _ignore) t))))
    (list :query query :preamble preamble :preamble-case-fold preamble-case-fold
	  :predicate predicate :action action :narrow t)))
```
and having the `pcase` clause in the preprocess stage be:

```elisp
(`(ancestors ,query) `(ancestors ,(org-ql--pre-process-ancestors-query (rec query))))
```

but I'm having an issue getting this to work with `org-ql--query-predicate`. It seems like it's not byte-compiling this new preprocessed query for some reason. I'll keep looking and report anything back.

Anyway, thanks again!

mm-- commented:
Oh, now I see that my issue was I wasn't quoting a list during macroexpansion, causing the byte-compilation to give a warning that the `:query` keyword was being called as a function, thus triggering the signal for `Invalid Org QL query`.

mm-- commented:
Okay, I've added another commit that uses your original `ancestors` code (much cleaner by the way), but moves the byte compiling of the inner query to the preprocess stage. It also fixes the slight issue with the `(parent (children (etc)))` I mentioned earlier.

I haven't used `org-ql--value-at` (I'm just using `org-ql--select-cached`) so maybe there's more performance issues that I'm missing.

mm-- commented:
Sorry for the spam. This latest commit uses `org-ql--value-at` with the compiled predicate function of the inner query.

alphapapa commented on org-ql.el:
```diff
@@ -519,6 +519,13 @@ Or, when possible, fix the problem."
 		  (org-ql--sanity-check-form (cdr elem)))
 	     else do (check elem))))
 
+(defun org-ql--inner-query-predicate (preprocessedquery)
```
What's the purpose of this function?  Preambles won't be used for these ancestor/parent "sub-queries," so why is `--query-preamble` being called?

mm-- replied:
My mistake. It looked like `--query-preamble` might have made some changes to the query in some way that might be needed. Removed this function.

alphapapa commented on org-ql.el:
```diff
@@ -921,6 +932,7 @@ Arguments STRING, POS, FILL, and LEVEL are according to
   "Return non-nil if current entry has children matching QUERY."
   (save-excursion
     (save-restriction
+      (widen) ;; Needed if we're inside an "ancestors" query
```
Since `widen` has been added, we could just use `org-with-wide-buffer`, which does all 3 of these operations.

mm-- replied:
Sure thing. Replaced in both functions.

alphapapa commented on org-ql.el:
```diff
@@ -938,10 +950,20 @@ Arguments STRING, POS, FILL, and LEVEL are according to
               :action (lambda ()
                         (throw 'found t)))))))))
 
+(org-ql--defpred parent (querypredicatefn)
```
Let's name the variable something simpler, like `predicate`.  :)

mm-- replied:
Done

alphapapa commented on org-ql.el:
```diff
@@ -938,10 +950,20 @@ Arguments STRING, POS, FILL, and LEVEL are according to
                         (throw 'found t)))))))))
 
+(org-ql--defpred parent (querypredicatefn)
+  "Return non-nil if the current entry's parent matches the compiled QUERYPREDICATEFN."
```
I think the fact that the predicate will be byte-compiled is an implementation detail that shouldn't matter to this function.

mm-- replied:
Ah, you're right. Updated the docstring.

alphapapa commented on org-ql.el:
```diff
@@ -938,10 +950,20 @@ Arguments STRING, POS, FILL, and LEVEL are according to
+      (widen)
+      (when (org-up-heading-safe)
+	(--when-let (org-ql--value-at (point) querypredicatefn)
+	  (not (eq 'org-ql-nil it)))))))
```
`org-ql--value-at` does not return `org-ql-nil`.  That symbol is used by it internally.

mm-- replied:
Hmm. Somehow this doesn't work without checking for `org-ql-nil`. It looks like `org-ql--value-at` returns `new-value` which is set to `org-ql-nil` if the predicate returns `nil`.

mm-- replied:
Okay, I fixed an issue where `org-ql--value-at` will return `org-ql-nil` when there's a cache miss and the `fn` returns `nil`.

mm-- commented:
Okay, I've implemented the requested changes in the latest commit. I also fixed a minor issue where `org-ql--value-at` returns `org-ql-nil` during a cache miss.

mm-- commented:
Squashed and rebased onto master.

alphapapa commented:
1.  Please do not squash.  In this case, you've made a change to `--value-at` which, if actually necessary, would be a bug fix that should be separate from this PR.  Squashing makes it more difficult to separate the changes.
2.  Looking at the diff, I see changes in unrelated code, including hunks with incorrect indentation.

Please review the diffs carefully.

mm-- commented:
Sorry, unsquashed. The minor change to `--value-at` is needed for `parent` and `ancestors` to work properly. Should they be written so they deal with `--value-at` returning `org-ql-nil` for now?

Looks like the changes to unrelated code (and indentation) were from adding `widen` to `children` and `descendants`, which was later simplified to `org-with-wide-buffer` (as per your suggestion).
Also, by default I used tabs to indent, but it looks like spaces are used here, which I missed, so I `untabify`'d.

I also previously missed two locations for `org-with-wide-buffer` and changed that.

Thanks for your patience.

alphapapa commented:
Hi Josh,

Thanks for your patience again.  Some changes have been made, including to `--value-at` in your #78, so please rebase this PR on master when you have time.

mm-- commented:
Hey alphapapa,

Thanks again for the guidance. I rebased the branch onto master, but then noticed a couple whitespace formatting errors on my part. I've changed my tabs to spaces (so indentation appears correct in the diff). I also fixed `parent` and `ancestors` to also use `org-with-wide-buffer`.

Let me know if anything needs to be squashed or changed in any way. I'd be happy to do that.

Best,
Josh

alphapapa commented:
Hi again,

Thanks, this is looking good.  I just pushed some tests in a separate commit.  It seems to work well, and performance is not bad.  These new predicates enable some interesting queries, like:

```el
;; Search for subprojects
(org-ql-search (org-agenda-files)
  '(and (todo "PROJECT")
        (ancestors (todo "PROJECT"))))

;; Search for all subtasks of projects, grouped by parent heading
(org-ql-search (org-agenda-files)
  '(and (todo)
        (ancestors (todo "PROJECT")))
  :super-groups '((:auto-parent t)))

;; Search for direct top-level tasks of projects
(org-ql-search (org-agenda-files)
  '(and (todo)
        (parent (todo "PROJECT")))
  :super-groups '((:auto-parent t)))
```
Of course, all of those presume using a `PROJECT` keyword to define projects.  If one defines a project as any task which has an ancestor task, one could use queries like:
```el
;; Search for all subtasks of top-level projects, grouped by parent heading
(org-ql-search (org-agenda-files)
  '(and (todo)
        (ancestors (and (todo)
                        (not (parent)))))
  :super-groups '((:auto-parent t)))

;; Search for all subtasks of all projects, including subprojects, grouped by project
(org-ql-search (org-agenda-files)
  '(and (todo)
        (ancestors (todo)))
  :super-groups '((:auto-parent t)))
```
Some of these should probably be added to the examples in the docs too.  Do you have any good examples?  I feel like there must be more interesting examples that aren't immediately obvious.

A quick thought I had: I think `ancestors` should be named `ancestor` instead, because it returns non-nil if any ancestor matches, not more than one.  In the same way, of course, the `descendants` predicate should also be named in the singular, so I should change that sometime.  What do you think?

Thanks for your work on this.  After adding docs and examples, I think it will be ready to merge for 0.4.

mm-- commented:
I think the examples you provided already are good illustrations of what most people might use `ancestors` and `parent` for. It's really for querying about things that can't be inherited, like parent/ancestor TODO states, non-inheritable tags, scheduled/deadlines, etc.

I can think of some other complex examples, but maybe they're a bit contrived and might not have as broad an appeal.

```elisp
;; Show dangling "next" tasks (unscheduled "next" tasks in a project
;; that has no scheduled todo items)
(org-ql-search (org-agenda-files)
  '(and (todo "NEXT")
	(not (scheduled))
	(ancestors (and (tags-local "project")
			(not (descendants (and (todo)
					       (scheduled)))))))
  :title "Next tasks that should be scheduled")

;; Subtasks of upcoming deadline items
;; (Can also be done for scheduled items)
(org-ql-search (org-agenda-files)
  '(and (todo)
	(ancestors (and (not (done))
			(deadline auto))))
  :super-groups '((:auto-parent t)))

;; TODO items whose parent is already DONE, and should therefore be
;; either marked DONE or CANCELLED
(org-ql-search (org-agenda-files)
  '(and (todo)
	(parent (done)))
  :super-groups '((:auto-parent t)))
```

In regards to naming `ancestors` as `ancestor`, naming is always a hard problem for me. A good rule of thumb that I try to follow (when I remember) is that names should be narrow and consistent. I think `ancestors` is more consistent with `descendants`. Somebody that already uses `descendants` would probably expect `ancestors` over `ancestor`. So I'd suggest the plural for now since people probably already depend on `descendants`. It's also worth nothing that `children` is plural as well.

Of course, you're right about the short-circuiting aspect of these predicates in that they only match one item. I guess I kind of thought about the predicates as saying something about "searching ancestors" or "searching descendants/children", in that they'll look through all ancestors/descendants/children until something is found. In that sense the plural form is a good descriptor.

I wonder in the future if it might be useful to have separate singular/plural predicates like `child` vs `children`, where the latter returns some list of matching headings. For example, looking at something like `(count (children (todo)))`. It's probably not useful strictly for matching, but maybe it'd be useful for sorting. It'd be interesting to be able to sort projects by their earliest scheduled descendant.


Should I add the examples you listed to `examples.org`?

alphapapa commented:
Hi Josh,

While we're at it, let me ask you about some of the examples you gave, to make sure we're on the same page.  These do get confusingly recursive after a while.

    ;; Show dangling "next" tasks (unscheduled "next" tasks in a project
    ;; that has no scheduled todo items)
    (org-ql-search (org-agenda-files)
      '(and (todo "NEXT")
            (not (scheduled))
            (ancestors (and (tags-local "project")
                            (not (descendants (and (todo)
                                                   (scheduled)))))))
      :title "Next tasks that should be scheduled")

Why use `tags-local` inside `ancestors` when you could use `tags` to match with inheritance?

    ;; Subtasks of upcoming deadline items
    ;; (Can also be done for scheduled items)
    (org-ql-search (org-agenda-files)
      '(and (todo)
            (ancestors (and (not (done))
                            (deadline auto))))
      :super-groups '((:auto-parent t)))

This one is really interesting!  We should definitely include it in the examples.

    ;; TODO items whose parent is already DONE, and should therefore be
    ;; either marked DONE or CANCELLED
    (org-ql-search (org-agenda-files)
      '(and (todo)
            (parent (done)))
      :super-groups '((:auto-parent t)))

This one is very cool.  I wonder if we should add it to the default `org-ql-views` as `Dangling tasks` or something like that.  I see you used "dangling" for that other example, but I feel like the term might fit this one better, unless there's an even better term for it.  What do you think?

> I think ancestors is more consistent with descendants. Somebody that already uses descendants would probably expect ancestors over ancestor. So I'd suggest the plural for now since people probably already depend on descendants.  

Right, this late in the 0.4 "cycle," any name changes should be postponed.  So we should keep the plurals for now and consider renaming them to singulars later.

> It's also worth nothing that children is plural as well.

Good point.

> Of course, you're right about the short-circuiting aspect of these predicates in that they only match one item. I guess I kind of thought about the predicates as saying something about "searching ancestors" or "searching descendants/children", in that they'll look through all ancestors/descendants/children until something is found. In that sense the plural form is a good descriptor.

Yeah, that's a good point too.  It's kind of a tough call.

I think, ideally, we should go with singular forms.  Another reason is that `parent` is singular, and it wouldn't make sense to have a plural `parents`, because each heading only has one parent.

> I wonder in the future if it might be useful to have separate singular/plural predicates like child vs children, where the latter returns some list of matching headings. For example, looking at something like (count (children (todo))). It's probably not useful strictly for matching, but maybe it'd be useful for sorting. It'd be interesting to be able to sort projects by their earliest scheduled descendant.

Yes, that's a good idea.  One of my long-term plans is to have some kind of "recursive" query support, or a way for queries to return descendants of matching headings.  That might be a good way to do it.

> Should I add the examples you listed to examples.org?

I don't want to wear out my welcome here, because you've already been very patient and helpful.  Feel free to add the examples if you like, but if you don't, I'll take care of that before merging.

alphapapa commented:
I think that should do it!  Thanks for your work and your patience!

mm-- commented:
Great! :tada: 

Thanks for adding the examples, and sorry for my late reply, I've been in kind of a time crunch this past week.

> Why use tags-local inside ancestors when you could use tags to match with inheritance?

It's more of a quirk of the way I specifically structure projects. Some tags I excluded from tag inheritance basically because I didn't like seeing subtasks labeled as "`project`" in the agenda when they themselves weren't projects. I also just use to use the "any task with a subtask is a project" after reading about it in [Bernt Hansen's org config](http://doc.norang.ca/org-mode.html#Projects), but later when I used to use orgzly I needed a simpler way to search projects. Hence the non-inherited `project` tag.

>  I see you used "dangling" for that other example, but I feel like the term might fit this one better, unless there's an even better term for it. What do you think?

Actually, yeah, that sounds like a better usage for term. And I see that you did add it as a view. :grin:
Hopefully people find it helpful. 

> I don't want to wear out my welcome here, because you've already been very patient and helpful. Feel free to add the examples if you like, but if you don't, I'll take care of that before merging.

I'd say "happy to help", but you beat me to adding the documentation, views, examples, etc. :grinning:
Again, thanks for all your guidance and tireless work you put into this!

Best,
Josh
